programs/cltester/dwarf.cpp

```diff
@@ -928,8 +1176,221 @@ namespace dwarf
          parse_debug_info_unit(result, whole_s, unit_s, {s.pos, s.pos + unit_length});
          s.skip(unit_length);
       }
+      std::sort(result.subprograms.begin(), result.subprograms.end());
+      for (size_t pos = 0; pos < result.subprograms.size();)
+         pos = fill_parents(result, pos, pos + 1);
+   }
+
+   Elf64_Word add_str(std::vector<char>& strings, const char* s)
+   {
+      if (!s || !*s)
+         return size_t(0);
+      auto result = strings.size();
+      strings.insert(strings.end(), s, s + strlen(s) + 1);
+      return result;
    }
 
+   template <typename T>
+   struct fill_later
+   {
+      size_t* offset = nullptr;
+   };
+
+   struct attr_form_value
+   {
+      using value_type = std::variant<uint8_t, uint64_t, std::string, fill_later<uint64_t>>;
+
+      uint32_t attr = 0;
+      uint32_t form = 0;
+      value_type value;
+
+      auto key() const { return std::pair{attr, form}; }
+
+      friend bool operator<(const attr_form_value& a, const attr_form_value& b)
+      {
+         return a.key() < b.key();
+      }
+   };
+
+   struct die_pattern
+   {
+      uint32_t tag = 0;
+      bool has_children = false;
+      std::vector<attr_form_value> attrs;
+
+      auto key() const { return std::tuple{tag, has_children, attrs}; }
```

Comment:
I would use tie to avoid copying the vector.

Reply:
changed